### ISSUE TITLE
RSTP image rotation

### ISSUE THREAD

**Issue by isanval** [EliasKotlyar/Xiaomi-Dafang-Hacks#1237]
Hi @EliasKotlyar ,

First of all, nice work, thank you very much for sharing it with all of us :-)

Is there any way to rotate 90º or 270º the image? I know there is a flip option (180º), but it would be interesting also to being able to rotate 90º or 270º.

In case it is not possible, is it very hard to implement it? May you consider implement it, please?

Regards,

**Comment by cwiggs** [comment]
Looks like that flip option is set here: https://github.com/EliasKotlyar/Xiaomi-Dafang-Hacks/blob/master/firmware_mod/bin/setconf

If you run setconf with invalid settings it shows this:

```
/system/sdcard/bin/setconf: invalid option -- h
Invalid Argument ?
Usage /system/sdcard/bin/setconf -g -k KEY -v VALUE
Where k to set value, g to get the value
        'f' flip mode set to
                '1' -> flip on
                '0' -> flip off
        'n' night mode set to
                '1' -> night mode on
                '0' -> night mode off
        'b' set bitrate to VALUE
        'o' OSD text set to VALUE
        'c' OSD color set VALUE to
                '0' for White
                '1' for Black
                '2' for Red
                '3' for Green
                '4' for Blue
                '5' for Cyan
                '6' for Yellow
                '7' for Purple
        'x' OSD position Y pos is set to VALUE
        's' OSD size
        'e' OSD font name (ttf)
        'p' OSD space between char is set to VALUE (can be negative)
        'w' fixed text width (0 variable, 1 fixed)
        'm' motion sensitivity (0 to 4) -1 to deactivate motion
        'z' display a circle when motion detected (-1 deactivated, use the OSD color numbers)
        'r' set detection region (shall be: x0,y0,x1,y1) the server need to be restarted to take into account the new value
        't' set tracking on/off (detection region is not taken into account anymore)
        'u' set time before launching script after no motion (to restore camera position) -1 to deactivate
        'h' set hardware volume (from mic)
        'i' set software volume percentage (X will add X44 the data, from 0 to xxx, -1 to do nothing)
        'q' set set filter number (1 or 2, 0 no filter)
        'l' set set highpass filter on/off
        'a' set set aec filter on/off
        'd' set frame rate (shall be FrmRateNum,FrmRateDen (example: 25,1 to get 25 images per second)
Example: to set osd text: /system/sdcard/bin/setconf -k o -v OSDTEXT
         to get osd text: /system/sdcard/bin/setconf -g o
```

So it looks like right now the only option is to flip or not, if someone knows how to update setconf we might be able to add options other than flip or not.

**Comment by jmtatsch** [comment]
You would need add that feature to setconf and our adapted v4l2rtspserver-master https://github.com/Dafang-Hacks/Main/tree/master/v4l2rtspserver-master.
Doable but the last commits of it might be broken.

**Comment by isanval** [comment]
@cwiggs , yes, I already saw what you said about setconf, but as you point, there is no option for rotate the image.

I saw the setconf is in source code also: https://github.com/Dafang-Hacks/Main/blob/master/v4l2rtspserver-tools/setconf.cpp I will take a look to flip implementation... it should not be very dificult adapt it for 90º or 270º image rotation.

@jmtatsch , what do you mean with "last commits of it might be broken"?

**Comment by Vinestou** [comment]
Could we get an option in the menu? Like 'flip'?

**Comment by Cocco17** [comment]
Rotating option would be great! My camera is mounted sideways, thus I need to rotate the picture with 90° - Any new progress on this request?

**Comment by Cocco17** [comment]
No chance to add that function?

**Comment by gadget78** [comment]
the flip option did work few months back, but new GUI has removed the flip option...
as i have mine mounted on ceiling, thus upside down
where would we configure this in manually ?

**Comment by Cocco17** [comment]
Hey guys, 

just FYI - I found a way to rotate the stream of the camera into HomeKit by using the Plugin "Homebridge-Camera-FFMpeg" - This plugin gives you the ability to integrate command line features, such as "rotation" (check out this link for more information: [https://ostechnix.com/how-to-rotate-videos-using-ffmpeg-from-commandline/](url))

You just need to add this line to your config.json: 

        {
            "name": "Camera FFmpeg",
            "mqtt": "localhost",
            "usermqtt": "mqtt_user",
            "passmqtt": "mqtt_password",
            "topic": "Flur/Dafang/#",
            "cameras": [
                {
                    "name": "Flur",
                    "motion": true,
                    "doorbell": true,
                    "switches": true,
                    "videoConfig": {
                        "source": "-rtsp_transport tcp -i rtsp://xxx.xxx.xxx.xxx:8554/unicast",
                        "stillImageSource": "-rtsp_transport http -i rtsp://xxx.xxx.xxx.xxx:8554/unicast -vframes 1 -r 1",
                        "videoFilter": "transpose=2",
                        "maxFPS": 21,
                        "maxBitrate": 299,
                        "audio": true,
                        "vflip": false,
                        "hflip": false,
                        "debug": true
                    }
                }
            ],
            "platform": "Camera-ffmpeg"
        }

Do you think there is any chance add this functionality to the Dafang Plugin?

**Comment by Cocco17** [comment]
Hey guys! Any news/updates/ideas for this issue? 
Br and thanks in advance!

**Comment by alkissack** [comment]
Flip doesnt work with Wyze camera (NEOS)
